[WGSL] Add support for constructors with explicit type arguments

tadeuzagallo opened the pull request:
#### 81eb4d4bbe45337831bc1e9bef61185ab55d4ea2
<pre>
[WGSL] Add support for constructors with explicit type arguments
<a href="https://bugs.webkit.org/show_bug.cgi?id=254653">https://bugs.webkit.org/show_bug.cgi?id=254653</a>
rdar://problem/107358772

Reviewed by Myles C. Maxfield.

Add support explicitly providing type arguments for constructors, such as `vecN&lt;T&gt;(...)`

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::OverloadResolver):
(WGSL:: const):
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::calculateRank):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
(WGSL::OverloadResolver::resolve const const):
(WGSL::resolveOverloads):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/invalid/overload.wgsl: Added.
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/262325@main">https://commits.webkit.org/262325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4727af4ff0cf1288faeefd61a9928f035bbf504b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1245 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1279 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1114 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1348 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1255 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1157 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1905 "Failed to checkout and rebase branch from PR 12100") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1187 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1146 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/1905 "Failed to checkout and rebase branch from PR 12100") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1140 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1177 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/1905 "Failed to checkout and rebase branch from PR 12100") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1188 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1125 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1156 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/311 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1215 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -56,11 +56,18 @@ struct ViableOverload {
+
+        Type* type { nullptr };
+        Inferred inferred { Inferred::Yes };
+    };
```
Conceptually, WGSL builtin types are a finite flat set. Some of those types are spelled with `<` characters in them (e.g. `vec<f32, 4>` but the language doesn't actually support arbitrary substitution. So, I'm curious why you're implementing this infrastructure to handle an arbitrary case, rather than just enumerating all the builtin types, and treating them all as siblings.

tadeuzagallo replied:
I'm not sure what (or where) "enumerating all builtin types" would happen, we might have different overall pictures of how this should work, so I feel like I'm not quite seeing what you mean, but we can chat about it.

As for this code specifically, all it adds is an extra bit to distinguishing between `vecN(...)`, where element type is inferred, and `vec2<i32>(...)`, where it's explicit. However, as I was typing the answer I realized it shouldn't be necessary, and found a bug in `conversionRank` which is why it didn't work before I added it. I'll fix the underlying bug and update the patch.

tadeuzagallo committed:
[WGSL] Add support for constructors with explicit type arguments
https://bugs.webkit.org/show_bug.cgi?id=254653
rdar://problem/107358772

Reviewed by Myles C. Maxfield.

Add support explicitly providing type arguments for constructors, such as `vecN<T>(...)`

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::OverloadResolver):
(WGSL:: const):
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::calculateRank):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
(WGSL::OverloadResolver::resolve const const):
(WGSL::resolveOverloads):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/invalid/overload.wgsl: Added.
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/262325@main
